src/testengine.user.storagestate/StorageStateUserManagerModule.cs

```diff
@@ -0,0 +1,258 @@
+            var testSuiteDefinition = singleTestInstanceState.GetTestSuiteDefinition();
+            var logger = singleTestInstanceState.GetLogger();
+
+            var userConfig = testState.GetUserConfiguration(testSuiteDefinition.Persona);
+
+            if (userConfig == null)
+            {
+                logger.LogError("Cannot find user config for persona");
+                throw new InvalidOperationException();
+            }
+
+            var user = environmentVariable.GetVariable(userConfig.EmailKey);
+
+            if (string.IsNullOrEmpty(user))
+            {
+                logger.LogError("Email key for persona cannot be empty");
+                throw new InvalidOperationException();
+            }
+
+            if (!IsValidEmail(user))
+            {
+                logger.LogError($"Invalid email {userConfig.EmailKey} for {testSuiteDefinition.Persona}");
+                throw new UserInputException();
+            }
+
+            if (!Settings.ContainsKey("FileSystem") || Settings.ContainsKey("FileSystem") && !(Settings["FileSystem"] is IFileSystem))
+            {
+                throw new UserInputException("File system not provided");
+            }
+
+            var fileSystem = Settings["FileSystem"] as IFileSystem;
+
+            var userName = GetUserNameFromEmail(user);
+            Location = !Location.EndsWith($"-{userName}") ? Location += $"-{userName}" : Location;
+
+            if (!fileSystem.Exists(Location))
+            {
+                fileSystem.CreateDirectory(Location);
+            }
+
+            var started = DateTime.Now;
+
+            // Wait a minimum of a five minutes
+            var timeout = Math.Max(5 * 60000, testState.GetTimeout());
+
+            var state = new LoginState()
+            {
+                DesiredUrl = desiredUrl,
+                UserEmail = user,
+                CallbackErrorFound = async () =>
+                {
+                    await context.StorageStateAsync(new BrowserContextStorageStateOptions { Path = Path.Combine(Location, "state.json") });
+                },
+                CallbackDesiredUrlFound = async (match) =>
+                {
+                    await context.StorageStateAsync(new BrowserContextStorageStateOptions { Path = Path.Combine(Location, "state.json") });
+                },
+                Module = this
+            };
+
+            logger.LogDebug($"Waiting for {timeout} milliseconds for desired url");
+            while (DateTime.Now.Subtract(started).TotalMilliseconds < timeout && !state.FoundMatch && !state.IsError)
+            {
+                try
+                {
+                    foreach (var page in context.Pages)
+                    {
+                        var loginHelper = new PowerPlatformLogin();
+                        state.Page = page;
+
+                        await loginHelper.HandleCommonLoginState(state);
+                    }
+                }
+                catch
+                {
+
```

Comment:
should this emit any log saying waiting and retrying after some time

Reply:
Updated to LogError